src/xrGameSpy/GameSpy_BrowsersWrapper.cpp

```diff
@@ -0,0 +1,290 @@
+#include "stdafx.h"
+#include "GameSpy_BrowsersWrapper.h"
+#include "xrCommon/xr_array.h"
+#include "xrCore/Threading/ScopeLock.hpp"
+
+CGSUpdateStatusAccumulator::CGSUpdateStatusAccumulator(GSUpdateStatus s) { Register(s); }
+
+void CGSUpdateStatusAccumulator::Register(GSUpdateStatus s) { accumulator.push_back(s); }
+
+void CGSUpdateStatusAccumulator::Reset(GSUpdateStatus s)
+{
+    ScopeLock sl(&lock);
+    accumulator.clear();
+    Register(s);
+}
+
+GSUpdateStatus CGSUpdateStatusAccumulator::GetOptimistic() const
+{
+    ScopeLock sl(&lock);
+    GSUpdateStatus res = GSUpdateStatus::Unknown;
+    for (auto s : accumulator)
+    {
+        if (s < res)
+            res = s;
+
+        if (res == GSUpdateStatus::Success)
+            break;
+    }
+    return res;
+}
+
+GSUpdateStatus CGSUpdateStatusAccumulator::GetPessimistic() const
+{
+    ScopeLock sl(&lock);
+    GSUpdateStatus res = GSUpdateStatus::Success;
+    for (auto s : accumulator)
```

Comment:
Use auto only if it's `std::or_not_std::some_iterator_with_long_name` 🙂 
Otherwise, we should write the type explicitly.

Reply:
Is it a rule from coding style? If yes I think it's... not a good rule. The 'AAA' (Almost-Always-Auto) is often recognized as a recommended practice. It often prevents us from possible bugs, helps us to make less changes in code after changing types of variables, forces the compiler to generate more optimal code and so on. The main counter argument is that sometimes we can't understand the real type quickly while lookung through the code. But it's a not appliciable for this case - the class is quite simple and we can easily know the 'guessed' type.
Do you think that 'for (GSUpdateStatus s : accumulator)' looks better? Or maybe you'd prefer 'for (auto it=accumulator.begin(); it!=accumulator.end(); ++it)? I don't like both of these variants.

Reply:
> The main counter argument is that sometimes we can't understand the real type quickly while looking through the code.

Quick type understanding is critical and should be taken into account even in such small case.

> Do you think that 'for (GSUpdateStatus s : accumulator)' looks better?

So, it may look worse, but quick type understanding is more important. However, in my opinion, this variant actually look better.

Reply:
But if `auto` will help compiler to produce more effective optimizations then it's a dilemma 🙂

Reply:
In this case the type deducing is quite obvious (from assigning explicitely-typed variable 'res').